plugins/Validation/src/Validation.ts

```diff
@@ -1,223 +1,346 @@
+                        },
+                    ]);
+                };
+            }
+
+            if (nestedProp === 'required') {
+                return (message?: string) => {
+                    instance.validators.push({
+                        fn: (value => Array.isArray(value) ? value.length > 0 : !!value),
+                        path: cleanPath,
+                        required: true,
+                        message: message || 'This field is required',
+                        conditions,
+                    });
+                };
+            }
+
+            if (nestedProp === 'validate') {
+                return (nestedValidator: ValidateFn<any>, message?: string) => {
+                    instance.validators.push({
+                        fn: nestedValidator,
+                        path: cleanPath,
+                        message,
+                        conditions,
+                    });
+                };
+            }
+
+            return buildProxy(instance, [...path, nestedProp], state, conditions.slice(0));
+        },
+    });
+}
+
+type PathFields<S> = (keyof S)[] | ((s: State<S extends (infer T)[] ? T : S>) => boolean);
+
+interface PathValidator<S> {
+    valid(fields?: PathFields<S>): boolean;
+
+    required(): boolean;
+
+    errors(fields?: PathFields<S>): string[];
+}
+
+export function Validation<T>(input: State<T>): PathValidator<T> {
+    const [instance] = input.attach(ValidationId);
+
+    if (instance instanceof Error) {
+        throw new Error(`Forgot to run ValidationAttach()`);
     }
-    return {
-        id: PluginID,
-        init: () => new ValidationPluginInstance() as {}
+
+    if (!(instance instanceof ValidatorInstance)) {
+        throw new Error('Expected plugin to be of ValidatorInstance');
     }
+
+    return {
+        valid(fields: PathFields<T>): boolean {
+            return this.errors(fields).length === 0;
+        },
+        required(): boolean {
+            return instance.isRequired(input);
+        },
+        errors(fields: PathFields<T>): string[] {
+            if (fields === undefined) {
+                return instance.errors(input);
+            }
+
+            let errors: string[] = [];
+
+            if (typeof fields === 'function') {
+                if (isArrayState(input)) {
+                    input.forEach(item => {
+                        // @ts-ignore
+                        if (fields(item)) {
+                            errors = [...errors, ...instance.errors(item)];
+                        }
+                    });
+                }
+            } else {
+                for (const field of fields) {
+                    errors = [...errors, ...instance.errors(input.nested(field))];
+                }
+            }
+
+            return errors;
+        },
+    };
+}
+
+export function ValidationAttach<T>(state: State<T>, config: ((validator: NestedValidator<T, T>) => void)) {
```

Comment:
this is getting close to what I was thinking about. It would be still possible to achieve the same without proxy, but maybe it is the other story.

As we discussed, you need to move your code to the Validation2 plugin, because it breaks compatibility with the previous one. 

Also when you are ready. Add the plugin to docs and a code sample like for other plugins.

Reply:
yeah, i'll create a new PR once i'm closer to final state with a new plugin folder.

w/r/t object config vs proxy... take a look at the example validation i have in PR description, in particular the conditional behavior. i feel like trying to replicate that as a POJO would be incredibly verbose.

at least this latest version, the cost of the proxy is only in one place (on attach), so don't imagine there is perf loss going a different route.